FIX: correctly resolve field names of relations

mleutenegger opened the pull request:
Fixes an error where fields of relations could not be filtered against.

Resolves 1) and 3) in #265

emteknetnz commented:
@mleutenegger Thanks for submitting this PR.  Would it be possible to add a unit test or two to this PR to shows that the bug has been fixed.  Cheers

mleutenegger commented:
👍  I'll try and add some

mleutenegger commented:
@emteknetnz I have added tests for the resolving of relations. I stumbled upon another bug in the `getDBField()` function which prevented the definition of deeper relations (>1). I have fixed that one too and added two assertions for this.

Please let me know what you think!

emteknetnz commented on src/QueryFilter/DataObjectQueryFilter.php:
```diff
@@ -419,13 +418,23 @@ protected function getDBField($field)
+
+            $relationName = array_shift($relationNames);
             $class = $this->getDataObjectInstance()->getRelationClass($relationName);
+            while ($class && count($relationNames) > 0) {
```
Just wondering if it's possible to rewrite this so that it doesn't have to loop through all the relation names?  Is it possible to reverse this so that it uses array_pop instead?

dhensby replied:
If it is done this way, then it should use `&& !empty($relationNames)` rather than a count

mleutenegger replied:
I can reverse the array and use `array_pop` instead combined with a `break` statement (or a `do...while` loop). However, the changes would be purely cosmetic, as it seems to me like there is no other way except for traversing the entire relation tree to find the actual final class and field type.

@dhensby Thanks for correcting me on the `count`!

emteknetnz reviewed (changes requested):
Seems good, I've run this through a [cross module integration test](https://travis-ci.org/github/creative-commoners/silverstripe-installer/builds/733920953) which it passed, just wondering if we could make one bit on it a little more efficient. Cheers.

mleutenegger commented:
@emteknetnz I have changed the condition to use `empty()` and used a do ... while loop with an `array_reverse`. I have to correct my statement from earlier, `array_pop` is faster than `array_shift` 😅 

However, I cannot fix the error with composer 2 on my end...

mleutenegger commented:
@emteknetnz It seems like the composer issue is fixed, could you re-run the tests for me?

emteknetnz commented:
@mleutenegger sorry been quite busy at my end, realistically I won't have a chance to look at this for a few weeks, sorry about the delay

mleutenegger commented:
@emteknetnz @dhensby any news on this?
